### PR TITLE
Refactor profiling

### PR DESCRIPTION
**Pull request by rhodrin**
Introduces a new basic (default) profiling option. The current 'basic' profiling has been renamed 'advanced'. The new basic option saves a lot of time in code which requires repeated calls to operator.

Note that this iteration has a modification to `operator.py` (see from line 448) to allow the 'dummy list' to be provided as an empty list. There is possible a different (better?) way of doing this? But providing a non-empty list with 'dummy' values does not seem like a nice solution.

### CODE REVIEW

File: devito/operator.py [hunk 1]
```diff
@@ -447,8 +447,10 @@ def _profile_output(self, args):
             itershapes = [",".join(str(i) for i in its) for its in v.itershapes]
             if len(itershapes) > 1:
                 name = "%s<%s>" % (k, ",".join("<%s>" % i for i in itershapes))
-            else:
+            elif len(itershapes) > 0:
```
Comment (FabioLuporini) [inline devito/operator.py]: nitpicking: this can be == 0

Comment (rhodrin) [reply devito/operator.py]: Do you mean == 1 by the way?

Comment (FabioLuporini) [reply devito/operator.py]: yeah 😅

**Comment by FabioLuporini** [comment]
Ah, some tests are failing because they need the operation count (`ops`). For those, and only for those, can you set the profile level back to the original level? you have several options to do this, probably the cleanest one is to write a decorator similar in spirit to `silencio` (but for profiling) and stick it on top of the failing tests

**Comment by rhodrin** [comment]
Applying an appropriate fix results in error due to #658. 

Ideally that issue should be resolved prior to these changes being implemented but there are some work around options if we decide not to fix that first. Thoughts?

**Comment by FabioLuporini** [comment]
> Applying an appropriate fix results in error due to #658.

yes, agreed. I just re-read the issue. I'm not sure what the codepy patch is exactly supposed to do, but I think we should go ahead and propose it on codepy's github, and see how it goes. Said that, I'm ok with merging the changes in #675 first. I just merged master into #675. If green, we can proceed and merge it; we can revisit the 'profiling' thing mentioned in there afterwards.

Thoughts?

**Comment by rhodrin** [comment]
Sounds like a plan.
I believe with that in place this branch should go green also (after I update it).

**Comment by FabioLuporini** [comment]
nice, merged

**Pull request merged**